src/combinatorics/simple_graph/handshake.lean

```diff
@@ -0,0 +1,259 @@
+    rintro rfl, simp only [and_true, eq_self_iff_true],
+    by_cases hveq : v₁ = v₂,
+    { left, subst hveq, },
+    { right,
+      have he : e₁ = ⟦(v₁, v₂)⟧ := (sym2.elems_iff_eq hveq).mp ⟨hv₁, hv₂⟩,
+      subst e₁,
+      have h' := sym2.mem_other_spec' hv₂,
+      rw [sym2.eq_swap, sym2.congr_left] at h',
+      convert h'.symm, }, },
+  { intro h, cases h; subst d₁, refl, },
+end
+
+/-- For a given vertex `v`, the injective map from the incidence set at `v` to the darts there. --/
+def dart_from_incidence_set (v : V) : G.incidence_set v → G.dart := λ ee, ⟨v, ee⟩
+
+lemma dart_from_incidence_set_inj (v : V) : function.injective (G.dart_from_incidence_set v) :=
+by { dsimp only [dart_from_incidence_set], exact sigma_mk_injective }
+
+section degree_sum
+variables [fintype V] [decidable_eq V] [decidable_rel G.adj]
+
+lemma dart_vert_fiber_card_eq_degree (v : V) :
+  (univ.filter (λ d, G.dart_vert d = v)).card = G.degree v :=
+begin
+  have hh := card_image_of_injective univ (G.dart_from_incidence_set_inj v),
+  rw [finset.card_univ, card_incidence_set_eq_degree] at hh,
+  convert hh,
+  ext d,
+  simp only [mem_image, true_and, mem_filter, set_coe.exists, mem_univ, exists_prop_of_true],
+  split,
+  { rintro rfl, use G.dart_edge d, use G.dart_edge_mem_incidence_set d,
+    apply dart.ext; simp [dart_from_incidence_set, dart_vert, dart_edge], },
+  { rintro ⟨e, he, rfl⟩, simp [dart_from_incidence_set, dart_vert] },
+end
+
+lemma dart_card_eq_sum_degrees : fintype.card G.dart = ∑ v, G.degree v :=
+begin
+  convert_to (univ : finset G.dart).card = _,
+  rw @card_eq_sum_card_fiberwise _ _ _ G.dart_vert _ univ (λ d h, mem_univ _),
+  simp [G.dart_vert_fiber_card_eq_degree],
+end
+
+lemma dart_edge_fiber (d : G.dart) :
+  (univ.filter (λ (d' : G.dart), G.dart_edge d' = G.dart_edge d)) = {d, G.dart_rev d} :=
+begin
+  ext d',
+  simp only [true_and, mem_filter, mem_insert, mem_univ, mem_singleton],
+  exact G.dart_edge_eq_iff d' d,
+end
+
+lemma dart_edge_fiber_card_eq_2 (e : sym2 V) (h : e ∈ G.edge_set) :
+  (univ.filter (λ (d : G.dart), G.dart_edge d = e)).card = 2 :=
+begin
+  refine quotient.ind (λ p h, _) e h, cases p with v w,
+  let d : G.dart := ⟨v, ⟦(v, w)⟧, h, sym2.mk_has_mem _ _⟩,
+  convert_to _ = finset.card {d, G.dart_rev d},
+  { rw [card_insert_of_not_mem, card_singleton], simp [G.dart_rev_no_fixedpoints d], },
+  congr, apply G.dart_edge_fiber d,
+end
+
+lemma dart_card_eq_twice_card_edges : fintype.card G.dart = 2 * G.edge_finset.card :=
+begin
+  convert_to (univ : finset G.dart).card = _,
+  rw @card_eq_sum_card_fiberwise _ _ _ G.dart_edge _ G.edge_finset
+    (λ d h, by { rw mem_edge_finset, apply G.dart_edge_mem }),
+  rw @sum_congr _ _ G.edge_finset G.edge_finset
+    (λ e, (univ.filter (λ (x : G.dart), G.dart_edge x = e)).card) (λ e, 2) _ rfl,
+  simp [mul_comm],
+  intros e h, rw mem_edge_finset at h,
+  exact G.dart_edge_fiber_card_eq_2 e h,
+end
+
+/-- The degree-sum formula. -/
```

Comment:
I think it's worth adding the name `handshaking lemma` here as well (and possibly in the tags too)